### DISCUSSION TITLE
How to fix Route [login] not defined.

### DISCUSSION THREAD

**Discussion by altimusin** [filamentphp/filament#5226] [Help]
Hi All,

How to fix Route [login] not defined.

I'm using Filament Admin Breeze plugin (https://github.com/jeffgreco13/filament-breezy)

My login page setting in filament config file.
```
    'auth' => [
        'guard' => env('FILAMENT_AUTH_GUARD', 'web'),
        'pages' => [
            'login' => \JeffGreco13\FilamentBreezy\Http\Livewire\Auth\Login::class,
        ],
    ],
```

![image](https://user-images.githubusercontent.com/43429317/206966989-bc72fb4b-73bf-4bf6-88c7-49cae284416b.png)

Thank you.

**Comment by danharrin** [comment]
Laravel limitation. Add a redirect route like `Route::redirect('/laravel/login', '/login')->name('login')` so Laravel knows where to send `login` requests

**Reply by altimusin** [reply]
Thanks

**Reply by simaremare** [reply]
I would do something like this.

```php
Route::get('/login', function () {
    return redirect(route('filament.admin.auth.login'));
})->name('login');
```

Where `admin` is your panel `id` in the version 3.x.

**Reply by arsal381** [reply]
@simaremare How to do this when using multiple panels?

**Reply by clicktodev** [reply]
@arsal381 I had to explicitly set `->login()` on non default panels

**Reply by NaungYeHtet** [reply]
This works for me

```
Route::get('/login', function () {
    $prevUrl = url()->previous();

    if (! $prevUrl) {
        abort(404); // or redirect some where
    }

    $path = parse_url($prevUrl, PHP_URL_PATH);

    $panelId = explode('/', trim($path, '/'))[0];

    if (! in_array($panelId, array_keys(Filament::getPanels()))) {
        abort(404);
    }

    return redirect(route("filament.{$panelId}.auth.login"));
})->name('login');
```

**Reply by rickysut** [reply]
just add ->login() at xxxxPanelProvider

**Reply by ronaldo-lipata** [reply]
> This works for me
> 
> ```
> Route::get('/login', function () {
>     $prevUrl = url()->previous();
> 
>     if (! $prevUrl) {
>         abort(404); // or redirect some where
>     }
> 
>     $path = parse_url($prevUrl, PHP_URL_PATH);
> 
>     $panelId = explode('/', trim($path, '/'))[0];
> 
>     if (! in_array($panelId, array_keys(Filament::getPanels()))) {
>         abort(404);
>     }
> 
>     return redirect(route("filament.{$panelId}.auth.login"));
> })->name('login');
> ```

Thank you!

**Comment by Loukili-Rachid** [comment]
i resolved that with replace the route('login') with filament route in the Authenticate.php middleware

```
    protected function redirectTo(Request $request): ?string
    {
        return $request->expectsJson() ? null : route('filament.admin.auth.login');
    }
```

**Reply by yosefmiller** [reply]
Can automatically generate that route, so that it can be used across multiple panels:

    protected function redirectTo(Request $request): ?string
    {
        return $request->expectsJson() ? null : Filament::getCurrentPanel()->route('auth.login');
    }

**Comment by ghost** [comment]
In my case I set my filament path to `/` because I use it on a subdomain.

So I am using this as a fix:

```php
Route::redirect('/login-redirect', '/login')->name('login');
```

Seems to work well.

**Comment by anasred** [comment]
in you PanelAdminProvider add this:

```php
return $panel
            ->default()
            ->id('admin')
            ->path('/')
            ->login(\Filament\Pages\Auth\Login::class) // -> This line will fix the issue
```


Note:
You have to delete any `/login` or `/` routes on your routes/web.php file

**Comment by Ramazanonat** [comment]
While expanding the API, I get this error. When I send a request with an incorrect token, it redirects to the login. It gets fixed by adding "accept application/json" in the header of the request, but I need to define and apply extra middleware for this, I'm looking for a more effective solution.

**Reply by danharrin** [reply]
Wdym expanding the API? Filament doesn't have an API

**Reply by Ramazanonat** [reply]
actually I'm getting this error while developing an API with Laravel. sorry for my english :(

**Comment by sinnbeck** [comment]
Laravel has a hidden feature for this. Just add this to a service providers boot method

```php
Authenticate::redirectUsing(fn() => Filament::getCurrentPanel()->route('auth.login'));

**Reply by ohchiko** [reply]
Worked for me, but with a little more addition and I did have multiple panels.

```php
use Filament\Facades\Filament;
use Filament\Http\Middleware\Authenticate;
use Illuminate\Auth\AuthenticationException;
use Illuminate\Support\ServiceProvider;

// ...

class AppServiceProvider extends ServiceProvider
{

    // ...

    public function boot(): void
    {
        Authenticate::redirectUsing(fn(): string => Filament::getLoginUrl());
        AuthenticateSession::redirectUsing(
            fn(): string => Filament::getLoginUrl()
        );
        AuthenticationException::redirectUsing(
            fn(): string => Filament::getLoginUrl()
        );
    }

    // ...

}
```

**Comment by arnoldski** [comment]
In order to solve it using Laravel 11 we just need to override this behaviour.
To do so, we need to modify the file bootstrap/app.php and add the following code (line 14):
```
<?php
 
use IlluminateFoundationApplication;
use IlluminateFoundationConfigurationExceptions;
use IlluminateFoundationConfigurationMiddleware;
 
return Application::configure(basePath: dirname(__DIR__))
    ->withRouting(
        web: __DIR__.'/../routes/web.php',
        commands: __DIR__.'/../routes/console.php',
        health: '/up',
    )
    ->withMiddleware(function (Middleware $middleware) {
        $middleware->redirectGuestsTo('login');
    })
    ->withExceptions(function (Exceptions $exceptions) {
        //
    })->create();
```

You can see we have added a new line inside the callback function that is passed as a parameter to the withMiddleware method:
`$middleware->redirectGuestsTo('login');`

**Reply by michaelthedev** [reply]
Can a condition be added to handle different login pages?

**Comment by johncarter-** [comment]
I'm using Filament on the root domain and have resorted to this. I hope it helps someone, but I have a feeling it's a bit wonky.

```php
return Application::configure(basePath: dirname(__DIR__))
    ->withRouting(
        web: __DIR__ . '/../routes/web.php',
        commands: __DIR__ . '/../routes/console.php',
        health: '/up',
    )
    ->withMiddleware(function (Middleware $middleware) {

        // Add these lines: 👇

        $middleware->redirectGuestsTo('/login');

        $middleware->redirectTo(
            guests: '/login',
            users: '/bookings'
        );
    })
    ->withExceptions(function (Exceptions $exceptions) {
        //
    })->create();
```

**Comment by MansourM** [comment]
I was also having this problem, this seems to be the best approach IMO
in laravel 11 `bootstrap/app.php` (or kernel in previous versions where you can define aliases)
override the `auth` alias and set it to the filament one
```php
->withMiddleware(function (Middleware $middleware) {
    $middleware->alias([
        'auth' => \Filament\Http\Middleware\Authenticate::class
    ]);
})
```
@danharrin i think it would be good to be done by fillament by default? or as an option
this error will always occur if somebody tries to add `auth` middleware to a page since (i also have custom login page, not sure if that plays a part too or not)

**Reply by danharrin** [reply]
Hi, no, it would be a breaking change and also would not be a sensible default where Filament is not the only part of the app that has auth (like if it is used as an admin panel aside the main app)

**Reply by MansourM** [reply]
Hello, thank you for your work on Filament.
I now understand why it was designed this way, and you’re absolutely right. However, since most apps won’t require multiple authentication setups, it might be helpful to include an optional helper function in the panel builder, or mention this in the documentation. I think that would be a nice addition.

**Reply by Infinitie04** [reply]
Thanks bro, its work for me 😄

**Comment by freshleafmedia** [comment]
This is now fixed in #15041
